File: lib/py/telepact/internal/validation/ValidateBytes.py
#|
#|  Copyright The Telepact Authors
#|  SPDX-License-Identifier: Apache-2.0
#|

from typing import TYPE_CHECKING, cast
import base64

from ...internal.types.TBytes import _BYTES_NAME

if TYPE_CHECKING:
    from .ValidationFailure import ValidationFailure
    from .ValidateContext import ValidateContext


def validate_bytes(value: object, ctx: 'ValidateContext') -> list['ValidationFailure']:
    from .GetTypeUnexpectedValidationFailure import get_type_unexpected_validation_failure

    if isinstance(value, bytes):
        if ctx.coerce_base64:
            set_coerced_path(ctx.path, ctx.base64_coercions)
        return []
    if isinstance(value, str):
        try:
            base64.b64decode(value)
            if not ctx.coerce_base64:
                set_coerced_path(ctx.path, ctx.bytes_coercions)
            return []
        except Exception as e:
            return get_type_unexpected_validation_failure([], value, 'Base64String')
    else:
        return get_type_unexpected_validation_failure([], value, _BYTES_NAME)
        

def set_coerced_path(path: list[str], coerced_path: dict[str, object]):
    part = path[0]

    if len(path) > 1:
        coerced_path[part] = coerced_path.get(part, {})
        set_coerced_path(path[1:], cast(dict[str, object], coerced_path[part]))
    else:
        coerced_path[part] = True

File: lib/ts/src/internal/validation/ValidateBytes.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { bytesName } from '../../internal/types/TBytes.js';
import { ValidationFailure } from './ValidationFailure.js';
import { ValidateContext } from './ValidateContext.js';
import { getTypeUnexpectedValidationFailure } from './GetTypeUnexpectedValidationFailure.js';
import { decodeBase64, encodeBase64 } from '../binary/Base64Util.js';

export function validateBytes(value: unknown, ctx: ValidateContext): ValidationFailure[] {
    if (value instanceof Uint8Array) {
        if (ctx.coerceBase64) {
            setCoercedPath(ctx.path, ctx.base64Coercions);
        }
        return [];
    }
    if (typeof value === 'string') {
        try {
            const decoded = decodeBase64(value);
            const reEncoded = encodeBase64(decoded);
            if (reEncoded !== value) {
                throw Error('Invalid base64')
            }
            if (!ctx.coerceBase64) {
                setCoercedPath(ctx.path, ctx.bytesCoercions);
            }
            return [];
        } catch (e) {
            return getTypeUnexpectedValidationFailure([], value, 'Base64String');
        }
    } else {
        return getTypeUnexpectedValidationFailure([], value, bytesName);
    }
}

function setCoercedPath(path: string[], coercedPath: Record<string, unknown>): void {
    const part = path[0];

    if (path.length > 1) {
        coercedPath[part] = coercedPath[part] || {};
        setCoercedPath(path.slice(1), coercedPath[part] as Record<string, unknown>);
    } else {
        coercedPath[part] = true;
    }
}